### DISCUSSION TITLE
Version 2.2.8: The disk space usage is too large

### DISCUSSION THREAD

**Discussion by ingale726** [milvus-io/milvus#24051] [Q&A and General discussion]
In the current situation, I upgraded to version 2.2.8 and set aside 500G space to insert 110,000,000 128 dimensional vectors.  After inserting about 50,000,000, the hard disk was full.  Later, I went back to version 2.2.3 with the same 500G space, and I predicted that there would be 300G space left。
I only tested it once,the situation affected me a lot

**Comment by xiaofan-luan** [comment]
> In the current situation, I upgraded to version 2.2.8 and set aside 500G space to insert 110,000,000 128 dimensional vectors. After inserting about 50,000,000, the hard disk was full. Later, I went back to version 2.2.3 with the same 500G space, and I predicted that there would be 300G space left。 I only tested it once,the situation affected me a lot

128 dim  110,000,000 data takes more than 660GB data. (index + data ) * 3. I'm not sure 500GB is the raw disk or S3, If it's raw disk I feels that 500G is enough.

**Reply by xiaofan-luan** [reply]
because if use minio, it takes at least 3 replica. data and index takes another 2X, and if you have kafka it is another 3X

**Reply by ingale726** [reply]
The server uses SSDS

**Reply by xiaofan-luan** [reply]
> The server uses SSDS

Minio itself uses 3X replicas. And index and collection doubles the space

**Comment by ingale726** [comment]
I will continue testing, but I won't be upgrading to 2.2.8 for a while

**Reply by xiaofan-luan** [reply]
> I will continue testing, but I won't be upgrading to 2.2.8 for a while

There should be no difference on data volume usage between 2.2.X to 2.2.8. One possible reason is the more active compaction.

Few things to check:
1) change the data retention time to smaller value, by default it's one day, change to 1 hour may help reduce the disk usage
2) don't do flush manually, avoid frequent flush cause small segments. 
3) Host 110,000,000 128 dimensional vectors should be really tight, did you check previous version is doable?

**Reply by xiaofan-luan** [reply]
So I would expect at least 300-400 GB for 110,000,000 128 dimensional vectors.
If compaction happens a lot, then 500GB might be not enough?

**Reply by ingale726** [reply]
I went back to version 2.2.3 and reinserted 110 million 128 dimensional vectors, which took up less than 100 gigabytes of space, compared to at least 400 gigabytes with version 2.2.8
I am currently building a 2.2.8 standalone version to recreate this problem
I'm in the middle of testing

**Reply by ingale726** [reply]
Nothing has changed in the code that creates collections and indexes
When I used 2.2.3 and 2.2.8 respectively

**Reply by yanliang567** [reply]
@xiao12mm how did you deploy milvus, helm or docker compose? did you change any default configurations?

**Reply by ingale726** [reply]
datanode segment:2048
pulsar     broker.conf
subscriptionExpirationTimeMinutes=10

**Reply by xiaofan-luan** [reply]
> datanode segment:2048 pulsar broker.conf subscriptionExpirationTimeMinutes=10

So when you talk about disk usage, is this refer to message queue usage or S3 usage?

**Reply by ingale726** [reply]
The server's SSD

**Reply by intothephone** [reply]
Same issue, only inserted 8w vectors, but the disk usage is 185G, unbelievable... following is the detailed info:

milvus version: 2.2.4
vector dimension: 768
number of vectors: 88199
index type: ivf_sq8
metirc type: ip
nlist: 8192

I'm using the default yaml in docker-compose:

standalone:
    container_name: milvus-standalone
    image: milvusdb/milvus:v2.2.4
    command: ["milvus", "run", "standalone"]

**Reply by xiaofan-luan** [reply]
Anyone can give me a ls -ltr so I know which is the directory take the most disk?

**Reply by ingale726** [reply]
![image](https://github.com/milvus-io/milvus/assets/42893941/1720a2a2-a44e-499a-8b72-481b8355d81e)
This is what it looks like in normal times

**Reply by ingale726** [reply]
I am now back to 2.2.3, the condition is normal

**Reply by ouyangxiu12345** [reply]
I have encountered the same problem. 240,000 vectors with 512 dimensions, which occupy 50G disk space. Today it has increased to 62G, while the number of vectors only increased by less than 10,000. I wrote a script to monitor the data and found that every 6 hours or so, both the index_files and insert_log will increase by around 1G.

**Reply by xiaofan-luan** [reply]
> I have encountered the same problem. 240,000 vectors with 512 dimensions, which occupy 50G disk space. Today it has increased to 62G, while the number of vectors only increased by less than 10,000. I wrote a script to monitor the data and found that every 6 hours or so, both the index_files and insert_log will increase by around 1G.

Compaction may consume extra disk space, a few extra gigabytes consumption is under expectation. If you tune the gc parameters, you should see the disk usage stable at certain level.

If anyone has datacoord and indexcoord log that would be really appreciated. So far we received feedback but it's not reproducible in our enviroment

**Reply by intothephone** [reply]
back to 2.2.3, inserted the same number of vectors: 88199, the disk usage is 1.3G, comparing to 185G of version 2.2.4...something must be wrong

**Comment by ingale726** [comment]
I think this happened after the upgrade. I did a test and installed version 2.2.8 directly. There didn't seem to be too much hard drive usage.

I upgraded from 2.2.3 to 2.2.8, and I did experience heavy hard drive usage

**Reply by xiaofan-luan** [reply]
> I think this happened after the upgrade. I did a test and installed version 2.2.8 directly. There didn't seem to be too much hard drive usage.
> 
> I upgraded from 2.2.3 to 2.2.8, and I did experience heavy hard drive usage

Good. suggestion. We will see issue on upgradeing from 2.2.3 to 2.2.6 or above. We will try 2.2.3 to upgrade to 2.2.8

**Reply by yanliang567** [reply]
@xiao12mm if you wait for 12 hours, does the disk usage goes down or just keep

**Reply by ingale726** [reply]
When the data is being written, the hard disk occupies a lot. After the data is written, the hard disk will restore some space. My feelings

**Comment by ruixiu** [comment]
I also encountered a situation where I didn't worry about the disk space being occupied when using versions before 2.2.8. After upgrading to version 2.2.8, the disk quickly became full, and I initially thought it was due to our large data volume, so I expanded the disk. However, after less than a day, the disk was once again occupied. Surprisingly, after restarting the Docker container of Milvus, the occupied disk was miraculously released and quickly filled up again. I haven't found a solution yet.

**Reply by xiaofan-luan** [reply]
> I also encountered a situation where I didn't worry about the disk space being occupied when using versions before 2.2.8. After upgrading to version 2.2.8, the disk quickly became full, and I initially thought it was due to our large data volume, so I expanded the disk. However, after less than a day, the disk was once again occupied. Surprisingly, after restarting the Docker container of Milvus, the occupied disk was miraculously released and quickly filled up again. I haven't found a solution yet.

Hi ruixiu,
What version you are using? will 2.2.9 solve your issue? 
As far as I know, if you didn't write data, milvus will consume around 10GB disk for the default retention params.
If you keep inserting data, you would expect rockdb occupies similar size of your insert rate * retention time + timetick

**Reply by ruixiu** [reply]
Upgraded to 2.2.9, but switched to a standalone version deployed using Docker Compose. Currently, the buffer/cache will continue to grow and the disk usage will continue to increase. After restarting the Docker container, it will be released.

**Comment by xiaofan-luan** [comment]
2.2.9 should fix a issue where compaction won't work. 
If you still hit the issue the quickest way to fix is to remove rocksmq dir

**Comment by hjlee9182** [comment]
@xiaofan-luan 
Hi, How do you calculate disk usage before inserting in milvus?
